aws_batch/pipeline.py

```diff
@@ -468,7 +468,10 @@ def return_merged_dict(dict1, dict2):
 
 def execute_command(command):
     print command
+    sys.stdout.flush()
     output = subprocess.check_output(command, shell=True)
+    print output
+    sys.stdout.flush()
     return output
 
 def wait_for_server(service_name, command, max_concurrent):
```

Comment:
1) I find it helpful to log messages whenever a wait for a resource can happen.

2) May be the AWS queuing system can handle that wait?

3) Otherwise it's also possible to package the wait like so:

```
    with SharedResource(resource_type="gsnap instance", requestor=job, priority="high", max_wait=3600) as gsnap_instance:
        gsnap_instance.run(job)
```

where the `__enter__` and `__exit__` functions implementing this block functionality will do the appropriate things, including raise an exception if the max_wait is exceeded.

4) The running of the job too can be encapsulated in such a block, in a way that enforces timeouts, requires keepalives, and allows you to filter the subprocess output.   Usually this is done using  a few more python threads. 

An individual log listener could be as simple as a thread that just gets lines from the subprocess stdout and inspects those lines and/or puts some of them into a queue from which another thread consumes and tsprints into the main process output log, possibly adding annotations that indicate which subprocess they belong to.